File: apps/etl-clojure/src/pulso/loader/correlations.clj
(ns pulso.loader.correlations
  (:require [pulso.xml.transform :as transform]
            [pulso.loader.lookups :as lookups]
            [pulso.loader.batch :as batch]))

(defn make-batchers
  "Creates batchers for correlation parent and child tables.
   Uses a dedicated nested-record returning-batcher to avoid
   cross-batcher complexity with top-level records."
  [ds batch-size]
  {:correlation     (batch/make-returning-batcher ds "correlation"
                      [:type :source_id :device_id
                       :creation_date :start_date :end_date]
                      batch-size)
   :metadata        (batch/make-batcher ds "correlation_metadata"
                      [:correlation_id :key :value] batch-size)
   :nested-record   (batch/make-returning-batcher ds "record"
                      [:record_type_id :source_id :device_id :unit_id
                       :value :creation_date :start_date :end_date]
                      batch-size)
   :record-metadata (batch/make-batcher ds "record_metadata"
                      [:record_id :key :value] batch-size)
   :record-link     (batch/make-batcher ds "correlation_record"
                      [:correlation_id :record_id] batch-size)})

(defn- process-returned-nested-records!
  "Takes returned id+metadata pairs from nested-record returning-batcher
   and feeds record metadata and correlation_record links into their batchers."
  [batchers pairs]
  (when pairs
    (doseq [{:keys [id metadata]} pairs]
      (let [{:keys [corr-id record-metadata]} metadata]
        ;; Link correlation <-> record
        ((:add! (:record-link batchers)) [corr-id id])
        ;; Record metadata
        (doseq [m record-metadata]
          ((:add! (:record-metadata batchers)) [id (:key m) (:value m)]))))))

(defn- process-returned-correlations!
  "Takes returned id+children pairs from correlation returning-batcher
   and feeds metadata and nested records into their batchers."
  [ds batchers pairs]
  (when pairs
    (doseq [{:keys [id metadata]} pairs]
      (let [{:keys [corr-metadata records]} metadata]
        ;; Correlation metadata
        (doseq [m corr-metadata]
          ((:add! (:metadata batchers)) [id (:key m) (:value m)]))
        ;; Nested records: pre-resolved FKs are already in the record maps
        (doseq [record-map records]
          (let [record-type-id (lookups/ensure-record-type-id! ds (:type record-map))
                source-id      (lookups/ensure-source-id! ds (:source-name record-map) (:source-version record-map))
                device-id      (lookups/ensure-device-id! ds (:device record-map))
                unit-id        (lookups/ensure-unit-id! ds (:unit record-map))
                row            [record-type-id source-id device-id unit-id
                                (:value record-map) (:creation-date record-map)
                                (:start-date record-map) (:end-date record-map)]
                nested-pairs   ((:add! (:nested-record batchers))
                                row {:corr-id id :record-metadata (:metadata record-map)})]
            (process-returned-nested-records! batchers nested-pairs)))))))

(defn process!
  "Transforms a Correlation element and batches it with nested records."
  [ds batchers element]
  (let [data       (transform/correlation-element->map element)
        source-id  (lookups/ensure-source-id! ds (:source-name data) (:source-version data))
        device-id  (lookups/ensure-device-id! ds (:device data))
        row        [(:type data) source-id device-id
                    (:creation-date data) (:start-date data) (:end-date data)]
        children   {:corr-metadata (:metadata data)
                    :records       (:records data)}
        pairs      ((:add! (:correlation batchers)) row children)]
    (process-returned-correlations! ds batchers pairs)))

(defn flush!
  "Two-phase flush: correlation parents → nested records → metadata + links.
   No longer needs record-batchers parameter."
  [ds batchers]
  ;; Phase 1: flush correlation parents, process children
  (let [corr-pairs ((:flush! (:correlation batchers)))]
    (process-returned-correlations! ds batchers corr-pairs))
  ;; Phase 2: flush nested records, process their metadata + links
  (let [nested-pairs ((:flush! (:nested-record batchers)))]
    (process-returned-nested-records! batchers nested-pairs))
  ;; Phase 3: flush all leaf batchers
  ((:flush! (:metadata batchers)))
  ((:flush! (:record-metadata batchers)))
  ((:flush! (:record-link batchers))))

File: apps/etl-python/pulso/loader/correlations.py
from pulso.loader import lookups
from pulso.loader.batch import Batcher, ReturningBatcher
from pulso.xml import transform

CORRELATION_COLUMNS = ["type", "source_id", "device_id", "creation_date", "start_date", "end_date"]
RECORD_COLUMNS = [
    "record_type_id", "source_id", "device_id", "unit_id",
    "value", "creation_date", "start_date", "end_date",
]


def make_batchers(pool, batch_size):
    """A dedicated nested-record ReturningBatcher avoids cross-batcher
    complexity with top-level records.py."""
    return {
        "correlation": ReturningBatcher(pool, "correlation", CORRELATION_COLUMNS, batch_size),
        "metadata": Batcher(pool, "correlation_metadata", ["correlation_id", "key", "value"], batch_size),
        "nested_record": ReturningBatcher(pool, "record", RECORD_COLUMNS, batch_size),
        "record_metadata": Batcher(pool, "record_metadata", ["record_id", "key", "value"], batch_size),
        "record_link": Batcher(pool, "correlation_record", ["correlation_id", "record_id"], batch_size),
    }


def _process_returned_nested_records(batchers, pairs):
    if not pairs:
        return
    for pair in pairs:
        record_id = pair["id"]
        meta = pair["metadata"]
        batchers["record_link"].add((meta["corr_id"], record_id))
        for m in meta["record_metadata"]:
            batchers["record_metadata"].add((record_id, m["key"], m["value"]))


def _process_returned_correlations(pool, batchers, pairs):
    if not pairs:
        return
    for pair in pairs:
        corr_id = pair["id"]
        meta = pair["metadata"]
        for m in meta["corr_metadata"]:
            batchers["metadata"].add((corr_id, m["key"], m["value"]))
        for record_map in meta["records"]:
            record_type_id = lookups.ensure_record_type_id(pool, record_map["type"])
            source_id = lookups.ensure_source_id(pool, record_map["source_name"], record_map["source_version"])
            device_id = lookups.ensure_device_id(pool, record_map["device"])
            unit_id = lookups.ensure_unit_id(pool, record_map["unit"])
            row = (
                record_type_id, source_id, device_id, unit_id,
                record_map["value"], record_map["creation_date"],
                record_map["start_date"], record_map["end_date"],
            )
            nested_pairs = batchers["nested_record"].add(
                row, {"corr_id": corr_id, "record_metadata": record_map["metadata"]}
            )
            _process_returned_nested_records(batchers, nested_pairs)


def process(pool, batchers, element):
    data = transform.correlation_element_to_map(element)
    source_id = lookups.ensure_source_id(pool, data["source_name"], data["source_version"])
    device_id = lookups.ensure_device_id(pool, data["device"])
    row = (data["type"], source_id, device_id, data["creation_date"], data["start_date"], data["end_date"])
    children = {"corr_metadata": data["metadata"], "records": data["records"]}
    pairs = batchers["correlation"].add(row, children)
    _process_returned_correlations(pool, batchers, pairs)


def flush(pool, batchers):
    """Three-phase flush: correlation parents -> nested records -> metadata + links."""
    corr_pairs = batchers["correlation"].flush()
    _process_returned_correlations(pool, batchers, corr_pairs)

    nested_pairs = batchers["nested_record"].flush()
    _process_returned_nested_records(batchers, nested_pairs)

    batchers["metadata"].flush()
    batchers["record_metadata"].flush()
    batchers["record_link"].flush()
